compiler/src/scala/quoted/runtime/impl/QuoteMatcher.scala

```diff
@@ -125,62 +125,75 @@ object QuoteMatcher {
   private def withEnv[T](env: Env)(body: Env ?=> T): T = body(using env)
 
   /** Evaluate the the result of pattern matching against a quote pattern.
-   *  Implementation of the runtime of `QuoteMatching.{ExprMatch,TypeMatch}.unapply`.
+   *  Implementation of the runtime of `QuoteMatching.ExprMatch.unapply`.
    */
-  def treeMatch(scrutinee: Tree, pattern: Tree)(using Context): Option[Tuple] = {
-    def isTypeHoleDef(tree: Tree): Boolean =
-      tree match
-        case tree: TypeDef =>
-          tree.symbol.hasAnnotation(defn.QuotedRuntimePatterns_patternTypeAnnot)
-        case _ => false
-
-    def extractTypeHoles(pat: Tree): (Tree, List[Symbol]) =
-      pat match
-        case tpd.Inlined(_, Nil, pat2) => extractTypeHoles(pat2)
-        case tpd.Block(stats @ ((typeHole: TypeDef) :: _), expr) if isTypeHoleDef(typeHole) =>
-          val holes = stats.takeWhile(isTypeHoleDef).map(_.symbol)
-          val otherStats = stats.dropWhile(isTypeHoleDef)
-          (tpd.cpy.Block(pat)(otherStats, expr), holes)
-        case _ =>
-          (pat, Nil)
-
-    val (pat1, typeHoles) = extractTypeHoles(pattern)
+  def exprMatch(scrutinee: Tree, pattern: Tree)(using Context): Option[Tuple] = {
```

Comment:
what's the advantage of splitting treeMatch like this? It seems we end up with quite a few lines duplicated.

Reply:
I refactored the code to only keep `treeMatch` as before. But added some optimizations to avoid using the `TypeMap` when it is not going to do anything.